src/libraries/Common/src/System/Diagnostics/DiagnosticsHelper.cs

```diff
@@ -7,62 +7,241 @@ namespace System.Diagnostics
 {
     internal static class DiagnosticsHelper
     {
-        internal static bool CompareTags(IEnumerable<KeyValuePair<string, object?>>? tags1, IEnumerable<KeyValuePair<string, object?>>? tags2)
+        // This is similar to System.Linq ToArray. We are not using System.Linq here to avoid the dependency.
+        internal static KeyValuePair<string, object?>[]? ToArray(IEnumerable<KeyValuePair<string, object?>>? tags)
         {
-            if (tags1 == tags2)
+            if (tags is null)
+            {
+                return null;
+            }
+
+            KeyValuePair<string, object?>[]? array = null;
+            if (tags is ICollection<KeyValuePair<string, object?>> tagsCol)
+            {
+                array = new KeyValuePair<string, object?>[tagsCol.Count];
+                if (tagsCol is IList<KeyValuePair<string, object?>> secondList)
+                {
+                    for (int i = 0; i < tagsCol.Count; i++)
+                    {
+                        array[i] = secondList[i];
+                    }
+
+                    return array;
+                }
+            }
+
+            if (array is null)
+            {
+                int count = 0;
+                using (IEnumerator<KeyValuePair<string, object?>> enumerator = tags.GetEnumerator())
+                {
+                    while (enumerator.MoveNext())
+                    {
+                        count++;
+                    }
+                }
+
+                array = new KeyValuePair<string, object?>[count];
+            }
```

Comment:
We shouldn't enumerate the enumerable twice.  We don't know what kind of enumerable it is; it could be really expensive to enumerate.

Reply:
What do you suggest doing? allocate?

Reply:
> What do you suggest doing? allocate?

Since `Tags` is an `IEnumerable<T>`, can we just use `new List<T>(tags)` instead of doing all this custom work?

Reply:
I'll get rid of the ToArray helper method as you suggested in another comment.